tests/unit/test_hnsw_tiered.cpp

```diff
@@ -165,6 +174,81 @@ TYPED_TEST(HNSWTieredIndexTest, addVector) {
 
         // Cleanup.
         delete jobQ;
-        VecSimIndex_Free(tiered_index);
+    }
+}
+
+TYPED_TEST(HNSWTieredIndexTest, manageIndexOwnership) {
+    std::shared_ptr<VecSimAllocator> allocator = VecSimAllocator::newVecsimAllocator();
+
+    // Create TieredHNSW index instance with a mock queue.
+    for (auto is_multi : {true, false}) {
+        size_t dim = 4;
+        HNSWParams params = {.type = TypeParam::get_index_type(),
+                             .dim = dim,
+                             .metric = VecSimMetric_L2,
+                             .multi = is_multi};
+        auto *jobQ = new JobQueue();
+        auto *index_ctx = new IndexExtCtx();
+        size_t memory_ctx = 0;
+        TieredIndexParams tiered_params = {.jobQueue = jobQ,
+                                           .indexCtx = index_ctx,
+                                           .submitCb = submit_callback,
+                                           .memoryCtx = &memory_ctx,
+                                           .UpdateMemCb = update_mem_callback};
+        TieredHNSWParams tiered_hnsw_params = {.hnswParams = params, .tieredParams = tiered_params};
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            HNSWFactory::NewTieredIndex(&tiered_hnsw_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx->index_ref.reset(tiered_index);
+        EXPECT_EQ(index_ctx->index_ref.use_count(), 1);
+        size_t initial_mem = memory_ctx;
+
+        // Create a dummy job callback that insert one vector to the underline HNSW index.
+        auto dummy_job = [](AsyncJob *job) {
+            auto *my_index =
+                reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(job->index);
+            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+            size_t dim = 4;
+            TEST_DATA_T vector[dim];
+            GenerateVector<TEST_DATA_T>(vector, dim);
+            if (my_index->index->indexCapacity() == my_index->index->indexSize()) {
+                my_index->index->increaseCapacity();
+            }
+            my_index->index->addVector(vector, my_index->index->indexSize(), false);
+        };
+
+        AsyncJob job(tiered_index->allocator, HNSW_INSERT_VECTOR_JOB, dummy_job, tiered_index);
+
+        // Wrap this job with an array and submit the jobs to the queue.
+        tiered_index->submitSingleJob((AsyncJob *)&job);
+        tiered_index->submitSingleJob((AsyncJob *)&job);
+        ASSERT_EQ(jobQ->size(), 2);
+
+        // Execute the job from the queue asynchronously, delete the index in the meantime.
+        auto run_fn = [&jobQ]() {
+            if (auto temp_ref = jobQ->front().relatedIndex.lock()) {
+                EXPECT_EQ(jobQ->front().relatedIndex.use_count(), 2);
+                reinterpret_cast<AsyncJob *>(jobQ->front().job)->Execute(jobQ->front().job);
+            }
```

Comment:
consider having this definition as past of the mock

Reply:
I could, but line 30 is specific to this test. Also, this is not precisely the general main loop for the background thread, it is lighter to suit this test (the more general is this one - https://github.com/RedisAI/VectorSimilarity/blob/tiered_index_insert_job/tests/unit/test_hnsw_tiered.cpp#L321)

Reply:
OK. Maybe we should have both options as part of the mock for easier test writing in the future, or at least the more general one and specifics can be written for tests that need it?

Reply:
Sure... I'll have it in the other PR where I use the general thread routine